From 9c522060d6419631bab73470db9686698ed9953f Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Mon, 11 May 2020 10:48:32 +0800
Subject: [PATCH] driver: gpu: dcss: release crtc state structure when reset
 crtc

When run suspend/resume feature, there is a memory leak as below:
[<00000000ea97f10b>] kmem_cache_alloc_trace+0x280/0x300
[<000000004364ceb3>] drm_atomic_helper_setup_commit+0x88/0x420
[<000000003f218540>] dcss_drm_atomic_commit+0xb4/0x230
[<00000000de362ada>] drm_atomic_commit+0x54/0x60
[<000000005b613a86>] drm_atomic_helper_disable_all+0x194/0x1a8
[<0000000075f5f417>] drm_atomic_helper_suspend+0x70/0x118
[<00000000328f68ed>] drm_mode_config_helper_suspend+0x3c/0x78
[<00000000c8454b52>] imx_drm_suspend+0x20/0x30
......
[<0000000066acf6c9>] state_store+0x90/0x100

Because the crtc state structure is not released in function
dcss_crtc_reset when enter system suspend status.
Therefore, refer to the reset function of other parts like
drm_plane and drm_connecter, add function
__drm_atomic_helper_crtc_destroy_state() to release
crtc state structure for fixing the memory leak.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dcss/dcss-crtc.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/imx/dcss/dcss-crtc.c b/drivers/gpu/drm/imx/dcss/dcss-crtc.c
index 736e9ece1172..fde625710265 100644
--- a/drivers/gpu/drm/imx/dcss/dcss-crtc.c
+++ b/drivers/gpu/drm/imx/dcss/dcss-crtc.c
@@ -58,8 +58,7 @@ static void dcss_crtc_reset(struct drm_crtc *crtc)
 	struct imx_crtc_state *state;
 
 	if (crtc->state) {
-		if (crtc->state->mode_blob)
-			drm_property_blob_put(crtc->state->mode_blob);
+		__drm_atomic_helper_crtc_destroy_state(crtc->state);
 
 		state = to_imx_crtc_state(crtc->state);
 		memset(state, 0, sizeof(*state));
-- 
2.17.1

